lib/jwt/x5c_key_finder.rb

```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+require 'base64'
+require 'jwt/error'
+
+module JWT
+  # If the x5c header certificate chain can be validated by trusted root
+  # certificates, and none of the certificates are revoked, returns the public
+  # key from the first certificate.
+  # See https://tools.ietf.org/html/rfc7515#section-4.1.6
+  class X5cKeyFinder
+    def self.from(x5c_header_or_certificates, root_certificates, crls)
+      store = build_store(root_certificates, crls)
+      signing_certificate, *certificate_chain = parse_certificates(x5c_header_or_certificates)
+      store_context = OpenSSL::X509::StoreContext.new(store, signing_certificate, certificate_chain)
+
+      if store_context.verify
+        signing_certificate.public_key
+      else
+        error = "Certificate verification failed: #{store_context.error_string}."
+        error = "#{error} Certificate subject: #{store_context.current_cert.subject}." if store_context.current_cert
+
+        raise JWT::VerificationError, error
+      end
+    end
+
+    def self.parse_certificates(x5c_header_or_certificates)
+      if x5c_header_or_certificates.all? { |obj| obj.is_a?(OpenSSL::X509::Certificate) }
+        x5c_header_or_certificates
+      else
+        x5c_header_or_certificates.map do |encoded|
+          OpenSSL::X509::Certificate.new(::Base64.strict_decode64(encoded))
+        end
+      end
+    end
+    private_class_method :parse_certificates
```

Comment:
Instead of making the static methods private afterwards with `private_class_method` you could use`class << self` combined with a `private` block.

```
class X5cKeyFinder
  class << self
    def from(x5c_header_or_certificates, root_certificates, crls)
    end
    
    private

    def parse_certificates(x5c_header_or_certificates)
    end

    def build_store(root_certificates, crls)
    end
  end
end

```